src/libpmemobj/critnib.c

```diff
@@ -0,0 +1,462 @@
+	if (!c)
+		return 0;
+	c->root = &nullnode;
+	os_mutex_init(&c->mutex);
+	return c;
+}
+
+/*
+ * internal: delete_node -- recursively free (to malloc) a subtree
+ */
+static void
+delete_node(struct critnib_node *__restrict n)
+{
+	if (n == &nullnode)
+		return;
+
+	if (n->shift != ENDBIT)
+		for (int i = 0; i < SLNODES; i++) {
+			if (n->child[i])
+				delete_node(n->child[i]);
+		}
+	Free(n);
+}
+
+/*
+ * critnib_delete -- destroy and free a critnib struct
+ */
+void
+critnib_delete(struct critnib *c)
+{
+	if (c->root)
+		delete_node(c->root);
+	os_mutex_destroy(&c->mutex);
+	for (struct critnib_node *m = c->deleted_node; m; ) {
+		struct critnib_node *mm = m->child[0];
+		Free(m);
+		m = mm;
+	}
+	for (int i = 0; i < DELETED_LIFE; i++)
+		for (int j = 0; j < 2; j++) {
+			if (c->pending_dels[i][j])
+				Free(c->pending_dels[i][j]);
+		}
+	Free(c);
+}
+
+/*
+ * internal: free_node -- free (to internal pool, not malloc) a node.
+ *
+ * We cannot free them to malloc as a stalled reader thread may still walk
+ * through such nodes; it will notice the result being bogus but only after
+ * completing the walk, thus we need to ensure any freed nodes still point
+ * to within the critnib structure.
+ */
+static void
+free_node(struct critnib *__restrict c, struct critnib_node *__restrict n)
+{
+	if (!n)
+		return;
+	n->child[0] = c->deleted_node;
+	c->deleted_node = n;
+}
+
+/*
+ * internal: alloc_node -- allocate a node from our pool or from malloc
+ */
+static struct critnib_node *
+alloc_node(struct critnib *__restrict c)
+{
+	if (!c->deleted_node)
+		return Malloc(sizeof(struct critnib_node));
+	struct critnib_node *n = c->deleted_node;
+	c->deleted_node = n->child[0];
+	return n;
+}
+
+/*
+ * crinib_insert -- write a key:value pair to the critnib structure
+ *
+ * Returns:
+ *  • 0 on success
+ *  • EINVAL if such a key already exists
+ *  • ENOMEM if we're out of memory
+ *
+ * Takes a global write lock but doesn't stall any readers.
+ */
+int
+critnib_insert(struct critnib *c, uint64_t key, void *value)
+{
+	os_mutex_lock(&c->mutex);
+	struct critnib_node *k = alloc_node(c);
+	if (!k)
+		return UNLOCK, ENOMEM;
+	k->path = key;
+	k->shift = ENDBIT;
+	k->child[0] = value;
+
+	struct critnib_node *n = c->root;
+	if (n == &nullnode) {
+		c->root = k;
+		return UNLOCK, 0;
+	}
+
+	struct critnib_node **parent = &c->root, *prev = c->root;
+
+	while (n->shift != ENDBIT && (key & (~NIB << n->shift)) == n->path) {
+		prev = n;
+		parent = &n->child[(key >> n->shift) & NIB];
+		n = *parent;
+	}
+
+	if (n == &nullnode) {
+		n = prev;
+		util_atomic_store_explicit64(
+			(uint64_t *)&n->child[(key >> n->shift) & NIB],
+			(uint64_t)k, memory_order_release);
+		return UNLOCK, 0;
+	}
+
+	uint64_t at = n->path ^ key;
+	if (!at) {
+		ASSERTeq(n->shift, ENDBIT);
+		free_node(c, k);
+		/* fail instead of replacing */
+		return UNLOCK, EINVAL;
+	}
+	int32_t sh = util_mssb_index64(at) & ~(SLICE - 1);
+
+	struct critnib_node *m = alloc_node(c);
```

Comment:
If I'm understanding correctly, right now you are using one extra node to store a key and it's full key.
You could use pointer tagging to distinguish between values and nodes. It would save you one allocation per insert.